airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -155,6 +172,22 @@ def empty_props_with_pk_if_present():
         if local_properties:
             yield local_properties
 
+    @staticmethod
+    def _next_chunk_id(property_chunks: Mapping[int, PropertyChunk]) -> Optional[int]:
+        """
+        Figure out which chunk is going to be read next.
+        It should be the one with the least number of records read by the moment.
+        """
+        non_exhausted_chunks = {
+            # We skip chunks that have already attempted a sync before and do not have a next page
+            chunk_id: property_chunk.record_counter
+            for chunk_id, property_chunk in property_chunks.items()
+            if property_chunk.first_time or property_chunk.next_page
```

Comment:
Previously, we had a condition that checked if `records > 0`, which we used to indicate that this was the first sync attempt so we did not need to check for the next page token. However, this had the unintended consequence of looping infinitely if a stream did not have any records because it would be tricked into thinking it was still the first one. Replacing this check with a boolean field `first_time` for each chunk partition allows us to definitively tell that this is in fact the first time

Reply:
@brianjlai thanks for the thorough investigation and the fix, I really missed that case. Changes look good to me and now the CAT run does not produce that many api calls, so I'm going to publish and release this